File: week13/15/3_Wed_kth_weak_heapq.cpp
#include <queue>
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> kWeakestRows(vector<vector<int>>& mat, int k) {
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> Q;
        // priority_queue를 min heap으로 사용하는 방법
        
        for (int i = 0; i < mat.size(); i++) {
            int soldier = 0;
            while (soldier < mat[i].size() && mat[i][soldier] == 1) soldier++;
            Q.push(make_pair(soldier, i));
        }
        vector<int> res;        
        for (int i = 0; i < k; i++) {             
            res.push_back(Q.top().second);
            Q.pop();
        }
        return res;
    }
};

File: week13/15/3_Wed_kth_weak_heapq.py
import heapq
from typing import List

class Solution:
    def kWeakestRows(self, mat: List[List[int]], k: int) -> List[int]:
        h = []
        for i, row in enumerate(mat):
            s = 0
            while s < len(row) and row[s] == 1: 
                s += 1
            heapq.heappush(h, (s, i))

        res = []
        for i in range(k):
            weakest =  heapq.heappop(h)
            res.append(weakest[1])
        
        return res

        